internal/sync/secret_mapper_sync.go

```diff
@@ -0,0 +1,314 @@
+		},
+		"gcp": {
+			{to: "GCP_B64ENCODED_CREDENTIALS", from: Raw{source: "googlecredentialConfig-authEncodedJson"}},
+		},
+		"digitalocean": {
+			{to: "DIGITALOCEAN_ACCESS_TOKEN", from: Raw{source: "digitaloceancredentialConfig-accessToken"}},
+			{to: "DO_B64ENCODED_CREDENTIALS", from: B64{source: "digitaloceancredentialConfig-accessToken"}},
+		},
+	}
+
+	driverMapping = map[string]string{
+		"vsphere": "vmwarevsphere",
+	}
+)
+
+var (
+	missingKey    = "Credential keys missing: %w"
+	missingSource = "Rancher Credentials secret named %s was not located"
+)
+
+type convert interface {
+	validate(data map[string][]byte) error
+	convert(data map[string][]byte) string
+}
+
+// Mapping defines a mapping between a source and destination secret keys.
+type Mapping struct {
+	from convert
+	to   string
+}
+
+// Template is a structure for rendering a template as a secret value.
+type Template struct {
+	template string
+	sources  []string
+}
+
+func (t Template) validate(data map[string][]byte) error {
+	for _, key := range t.sources {
+		if _, found := data[key]; !found {
+			return fmt.Errorf("key not found: %s", key)
+		}
+	}
+
+	return nil
+}
+
+func (t Template) convert(data map[string][]byte) string {
+	var renderedTemplate bytes.Buffer
+
+	if err := t.validate(data); err != nil {
+		return ""
+	}
+
+	stringData := map[string]string{}
+	for k, v := range data {
+		stringData[k] = string(v)
+	}
+
+	if t, err := template.New("").Parse(t.template); err != nil {
+		return err.Error()
+	} else if err = t.Execute(&renderedTemplate, stringData); err != nil {
+		return err.Error()
+	}
+
+	return base64.StdEncoding.EncodeToString(renderedTemplate.Bytes())
+}
+
+// Raw is a structure for storing a secret key without encoding.
+type Raw struct {
+	source string
+}
+
+func (r Raw) validate(data map[string][]byte) error {
+	if _, found := data[r.source]; !found {
+		return fmt.Errorf("key not found: %s", r.source)
+	}
+
+	return nil
+}
+
+func (r Raw) convert(data map[string][]byte) string {
+	return string(data[r.source])
+}
+
+// B64 is a structure for encoding a secret key as base64.
+type B64 struct {
+	source string
+}
+
+func (r B64) validate(data map[string][]byte) error {
+	if _, found := data[r.source]; !found {
+		return fmt.Errorf("key not found: %s", r.source)
+	}
+
+	return nil
+}
+
+func (r B64) convert(data map[string][]byte) string {
+	return base64.StdEncoding.EncodeToString(data[r.source])
+}
+
+// SecretMapperSync is a structure mirroring variable secret state of the Rancher secret data.
+type SecretMapperSync struct {
+	*SecretSync
+	RancherSecret *corev1.Secret
+}
+
+// NewSecretMapperSync creates a new secret mapper object sync.
+func NewSecretMapperSync(cl client.Client, capiProvider *turtlesv1.CAPIProvider) Sync {
+	if capiProvider.Spec.Credentials == nil ||
+		capiProvider.Spec.Credentials.RancherCloudCredential == "" ||
+		knownProviderRequirements[capiProvider.Spec.Name] == nil {
+		return nil
```

Comment:
same as above, some debug logs maybe? Just to avoid cases of not knowing when nil silently returned

Reply:
The nil logic is handled as a normal case in sync interface. This just skips the component entirely for the reconciled CAPIProvider resource. This may generate many lines of unnecessary information, not sure about that. Maybe log.V(9) or something as a level?

Reply:
Fixed